File: src/com/suushiemaniac/cubing/bld/model/enumeration/puzzle/DodecahedronPuzzle.java
package com.suushiemaniac.cubing.bld.model.enumeration.puzzle;

import com.suushiemaniac.cubing.alglib.lang.MegaminxAlgorithmReader;
import com.suushiemaniac.cubing.alglib.lang.NotationReader;
import com.suushiemaniac.cubing.bld.analyze.BldPuzzle;
import com.suushiemaniac.cubing.bld.analyze.TwoBldCube;
import com.suushiemaniac.cubing.bld.model.enumeration.piece.PieceType;
import net.gnehzr.tnoodle.scrambles.Puzzle;
import puzzle.CubePuzzle;
import puzzle.MegaminxPuzzle;

import java.util.function.Supplier;

import static com.suushiemaniac.cubing.bld.model.enumeration.piece.DodecahedronPieceType.*;

public enum DodecahedronPuzzle implements TwistyPuzzle { //TODO correct puzzle implementations
	KILO(2, MegaminxPuzzle::new, new TwoBldCube(), CORNER),
	MEGA(2, () -> new CubePuzzle(2), new TwoBldCube(), CORNER, CENTER, EDGE),
	MASTERKILO(2, () -> new CubePuzzle(2), new TwoBldCube(), CORNER, CENTER, EDGE),
	GIGA(2, () -> new CubePuzzle(2), new TwoBldCube(), CORNER, CENTER, EDGE);

	public static DodecahedronPuzzle fromSize(int size) {
		for (DodecahedronPuzzle puzzle : values()) {
			if (puzzle.getSize() == size) {
				return puzzle;
			}
		}

		return MEGA;
	}

	private static NotationReader READER_INST = new MegaminxAlgorithmReader();

	private int size;
	private Supplier<Puzzle> scramblingPuzzleGen;
	private BldPuzzle analyzingPuzzle;
	private PieceType[] types;

	DodecahedronPuzzle(int size, Supplier<Puzzle> scramblingPuzzleGen, BldPuzzle analyzingPuzzle, PieceType... types) {
		this.size = size;
		this.scramblingPuzzleGen = scramblingPuzzleGen;
		this.analyzingPuzzle = analyzingPuzzle;
		this.types = types;
	}

	@Override
	public int getSize() {
		return this.size;
	}

	@Override
	public Puzzle getScramblingPuzzle() {
		return this.scramblingPuzzleGen.get();
	}

	@Override
	public Supplier<Puzzle> generateScramblingPuzzle() {
		return this.scramblingPuzzleGen;
	}

	@Override
	public BldPuzzle getAnalyzingPuzzle() {
		return this.analyzingPuzzle;
	}

	@Override
	public PieceType[] getPieceTypes() {
		return this.types;
	}

	@Override
	public NotationReader getReader() {
		return READER_INST;
	}
}

File: src/com/suushiemaniac/cubing/bld/model/enumeration/puzzle/DodecahedronPuzzle.kt
package com.suushiemaniac.cubing.bld.model.enumeration.puzzle

import com.suushiemaniac.cubing.alglib.lang.MegaminxAlgorithmReader
import com.suushiemaniac.cubing.alglib.lang.NotationReader
import com.suushiemaniac.cubing.bld.analyze.BldCube
import com.suushiemaniac.cubing.bld.analyze.BldPuzzle
import net.gnehzr.tnoodle.scrambles.Puzzle
import puzzle.CubePuzzle
import puzzle.MegaminxPuzzle

import com.suushiemaniac.cubing.bld.model.enumeration.piece.DodecahedronPieceType.*
import com.suushiemaniac.cubing.bld.model.enumeration.piece.PieceType

enum class DodecahedronPuzzle(override val size: Int, private val scramblingPuzzleGen: () -> Puzzle, vararg types: PieceType) : TwistyPuzzle { //TODO correct puzzle implementations
    KILO(2, ::MegaminxPuzzle, CORNER),
    MEGA(3, { CubePuzzle(2) }, CORNER, CENTER, EDGE),
    MASTERKILO(4, { CubePuzzle(2) }, CORNER, CENTER, EDGE),
    GIGA(5, { CubePuzzle(2) }, CORNER, CENTER, EDGE);

    override val pieceTypes: Array<out PieceType> = types

    override val analyzingPuzzle: BldPuzzle
        get() = BldCube(CubicPuzzle.THREE) // FIXME

    override val reader: NotationReader
        get() = READER_INST

    override fun supplyScramblingPuzzle(): () -> Puzzle {
        return this.scramblingPuzzleGen
    }

    companion object {
        fun fromSize(size: Int): DodecahedronPuzzle {
            for (puzzle in values()) {
                if (puzzle.size == size) {
                    return puzzle
                }
            }

            return MEGA
        }

        private val READER_INST = MegaminxAlgorithmReader()
    }
}
